Rustls support with dynamic choice of TLS provider for cloud

muzarski opened the pull request:
Bases on https://github.com/scylladb/scylla-rust-driver/pull/1190
Closes: https://github.com/scylladb/scylla-rust-driver/pull/1190
Fixes: https://github.com/scylladb/scylla-rust-driver/issues/771

What changes compared to the base PR:
- `__tls` feature is removed
- TLS provider for the cloud is not statically predetermined based on enabled feature set. Users are now able to choose the provider in runtime via `CloudSessionBuilder::new()`.
- renamed `cloud.rs` example to `cloud-openssl.rs`. Introduced analogous `cloud-rustls.rs` example.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [x] I have adjusted the documentation in `./docs/source/`.
- [ ] I added appropriate `Fixes:` annotations to PR description.

Lorak-mmk commented on scylla/src/cloud/config.rs:
```diff
@@ -92,30 +92,39 @@ impl CloudConfig {
                     SslVerifyMode::PEER
                 });
-                let ca = datacenter.openssl_ca();
+                let ca = datacenter.ca_cert.openssl_ca().expect(
+                    "Driver bug! User chose OpenSSL
+                    provider, but datacenter CA cert is not OpenSSL",
+                );
```
Is there no way to prevent this statically?

muzarski replied:
Not that I know of, unfortunately... 

My first idea is to hold certificates in both formats (openssl, rustls) - each hidden behind its corresponding feature flag. But I'm not a fan of it. The ### size of `Datacenter` structure will grow as we introduce new tls providers.

Other idea is to hold the certificate as `String`, and do the parsing to either `openssl::x509::X509` or `rustls::pki_types::CertificateDer` (in the corresponding match arm) here. Then we do not need to "expect" anything. But this has a drawback - in new version we need to do the cert parsing each time we are in `CloudConfig::make_tls_config_for_scylla_cloud_host`. Currently, we do that only once during `CloudSessionBuilder::new()`.

Lorak-mmk replied:
Ok I checked out the PR locally and understand the problem now.

Lorak-mmk commented on scylla/src/cloud/config.rs:
```diff
@@ -306,6 +313,34 @@ impl Datacenter {
+    Rustls023(rustls::pki_types::CertificateDer<'static>),
+}
+
+impl TlsCert {
+    #[cfg(feature = "openssl-010")]
+    fn openssl_ca(&self) -> Option<&openssl::x509::X509> {
+        match self {
+            TlsCert::OpenSsl010(ca) => Some(ca),
+            #[cfg(feature = "rustls-023")]
+            TlsCert::Rustls023(_) => None,
+        }
+    }
+
+    #[cfg(feature = "rustls-023")]
+    fn rustls_ca(&self) -> Option<&rustls::pki_types::CertificateDer<'static>> {
+        match self {
+            TlsCert::Rustls023(ca) => Some(ca),
+            #[cfg(feature = "openssl-010")]
+            TlsCert::OpenSsl010(_) => None,
+        }
+    }
+}
```
Those methods are intended to only ever return Some for a single variant of the enum.
When adding new variants, we will need to add match arms to every such method.
Why not use `if let`?

muzarski replied:
You are correct, obviously. I tend to forget that `if let` syntax works for any enum, not just `Option`...

Lorak-mmk reviewed (changes requested):
Looks nice!
I wonder if for 2.0/3.0 it would make more sense to use (sealed) traits for TLS-related data instead of enums.
It should reduce amount of conditional compilation flags, but at the same time worsen the discoverability in docs.

muzarski commented on scylla/src/cloud/config.rs:
```diff
@@ -114,6 +120,60 @@ impl CloudConfig {
+                    broadcast by the node was left as address to open connection to.",
+                host_id
+            );
+            // FIXME: Is this acceptable to do such fallback?
+            return Ok(untranslated_address);
+        };
+
+        let Some(dc_config) = self.get_datacenters().get(dc) else {
+            warn!( // FIXME: perhaps error! would fit here better?
+                "Datacenter {} that node {} resides in not found in the Cloud config; ; therefore address \
+                    broadcast by the node was left as address to open connection to.",
+                dc, host_id
+            );
+            // FIXME: Is this acceptable to do such fallback?
+            return Ok(untranslated_address);
+        };
```
I did not change anything in this piece of code. See the changes from original PR: https://github.com/scylladb/scylla-rust-driver/pull/1190/files#diff-1c77c9c969cb54248037a2a0b5836c362e1320a752786c408bda8718111297ffR144-R196

muzarski replied:
OK, so from my investigation, the fallbacks were here before. The code has just been rewritten to use `let ... else` syntax. We even get rid of one fallback, when `resolve_hostname` function fails.

Original code:
```rust
let hostname = dc_config.get_server();
if let Ok(resolved) = resolve_hostname(hostname).await {
    *address = NodeAddr::Untranslatable(resolved)
} else {
    warn!(
            "Couldn't resolve address: {} of datacenter {} that node {} resides in; therefore address \
             broadcast by the node was left as address to open connection to.",
            hostname, dc, host_id
        );
}

// The endpoint is returned later - unmodified.
```

New code:
```rust
let hostname = dc_config.get_server();
let resolved = resolve_hostname(hostname).await
    // inspect_err() is stable since 1.76.
    // TODO: use inspect_err once we bump MSRV to at least 1.76.
    .map_err(|err| {
        warn!(
            "Couldn't resolve address: {} of datacenter {} that node {} resides in; therefore address \
                broadcast by the node was left as address to open connection to.",
            hostname, dc, host_id
        );
        err
    })
    .map_err(Arc::new)
    .map_err(TranslationError::IoError)?;
```

muzarski commented on docs/source/connecting/connecting.md:
Well, we did not even have any non-cloud tests for openssl. In TLS workflow we just run the example, where we connect to the cluster using openssl. Analogous example is introduced and run in CI for rustls.

muzarski commented on scylla/src/cloud/config.rs:
```diff
@@ -109,6 +109,28 @@ impl CloudConfig {
+                root_store.add(datacenter.rustls_ca().clone())?;
+                let builder = ClientConfig::builder();
+                let builder = if datacenter.get_insecure_skip_tls_verify() {
+                    let supported = builder.crypto_provider().signature_verification_algorithms;
+                    builder
+                        .dangerous()
+                        .with_custom_certificate_verifier(Arc::new(NoCertificateVerification {
+                            supported,
+                        }))
```
I'd need to investigate to answer that. Maybe @wprzytula knows something about it.

wprzytula replied:
I know nothing more in this regard. This is work of @nrxus.

Lorak-mmk replied:
This is not really important for now anyway, right? Its just cloud, which is behind an unstable feature.
We will need to verify that before releasing it officially though.

muzarski commented:
Rebased on main (clippy fixes)

muzarski commented on scylla/src/cloud/mod.rs:
```diff
@@ -35,25 +32,29 @@ pub(crate) fn set_ssl_config_for_scylla_cloud_host(
+                builder.set_private_key(key)?;
+                let context = builder.build();
+                TlsContext::OpenSsl010(context)
+            }
```
I'll do it in a separate commit. This piece of code is moved around and modified quite often, so it's a pain to change earlier and rebase.

muzarski commented on scylla/src/network/connection.rs:
```diff
@@ -212,35 +208,59 @@ struct TaskResponse {
+    // Option 3: Serverless Cloud. The Option<TlsConfig> remains None in ConnectionConfig until it reaches
+    // NodeConnectionPool::new(). Inside that function, the field is mutated to contain TlsConfig specific
+    // for the particular node. (The TlsConfig must be different, because SNIs differ for different nodes.)
+    // Thenceforth, all connections to that node share the same TlsConfig.
```
I'm not sure, but I see that this docstring is completely modified later:
```
/// Encapsulates TLS-regarding configuration that is specific for a particular endpoint.
///
/// Both use cases are supported:
/// 1. User-provided global TlsContext. Then, the global TlsContext is simply cloned here.
/// 2. Serverless Cloud. Then the TlsContext is customized for the given endpoint,
///    and its SNI information is stored alongside.
```

The change is introduced in "connection: introduce TlsProvider".

muzarski commented:
v1.1: Addressed @Lorak-mmk comments:
- extracted `TlsInfo::get_dc_tls_context()` method
- using `if let ...` syntax in `TlsCert::[openssl/rustls]_ca` method.

muzarski commented:
>     * using `if let ...` syntax in `TlsCert::[openssl/rustls]_ca` method.

Ehhh... Now when we enable `unstable-cloud` and **only one** of `openss-x`/`rustls-x`, the compiler produces following warning:
```
 error: irrefutable `if let` pattern
   --> scylla/src/cloud/config.rs:331:12
    |
331 |         if let TlsCert::OpenSsl010(ca) = self {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this pattern will always match, so the `if let` is useless
    = help: consider replacing the `if let` with a `let`
    = note: `-D irrefutable-let-patterns` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(irrefutable_let_patterns)]`
```

muzarski commented:
Is adding `#[allow(irrefutable_let_patterns)]` acceptable here? @Lorak-mmk

Lorak-mmk commented:
> Is adding `#[allow(irrefutable_let_patterns)]` acceptable here? @Lorak-mmk

Yes

muzarski commented:
v1.2: Silenced `irrefutable_let_patterns` compiler warnings emitted when `cloud::config::TlsCert` consists of only one variant.

wprzytula commented on scylla/Cargo.toml:
```diff
@@ -15,7 +15,6 @@ all-features = true
-default = []
 openssl-010 = ["dep:tokio-openssl", "dep:openssl"]
 cloud = [
     "openssl-010",
```
It looks to have been introduced by @nrxus, probably by accident. Let's just fix it.

nrxus replied:
fwiw I don't see my original PR removing the `default = []` so it was probably just some mix up during re-basing after the fact: https://github.com/scylladb/scylla-rust-driver/pull/1182/commits/f0fad3bcade089ec63bf88603a4ffdeb65670d69#diff-8737b034c595447843934804c583979f35de74fd8964f7af9bb770e2bb9771a5

wprzytula commented on scylla/src/network/connection.rs:
```diff
@@ -1367,22 +1386,27 @@ impl Connection {
+                    return Ok(spawn_router_and_get_handle(
+                        config,
+                        stream,
+                        receiver,
```
I'm not at all convinced that it's worthy.

wprzytula commented on docs/source/connecting/connecting.md:
````diff
@@ -68,20 +68,27 @@ specify the secure connection bundle as follows:
 ```
 
+> ***Note***\
+> `CloudSessionBuilder::new()` accepts two parameters. One of them is path to the configuration file.
+> Second parameter is responsible for choosing the underlying TLS provider library.
````
:wrench:

wprzytula replied:
```suggestion
> `CloudSessionBuilder::new()` accepts two parameters. The first is a path to the configuration file.
> The second parameter is responsible for choosing the underlying TLS provider library.
```

muzarski commented:
v1.3: Reverted `default = []` in `[features]` section of `scylla/Cargo.toml`.

wprzytula commented on .github/workflows/serverless.yaml:
```diff
@@ -31,6 +32,14 @@ jobs:
         run: rustup update
       - name: Check
         run: cargo check --verbose
+      # Cloud-related feature sets.   
+      - name: Cargo check with unstable-cloud and openssl-x features
+        run: cargo check --all-targets --manifest-path "scylla/Cargo.toml" --features "unstable-cloud" --features "openssl-010"
+      - name: Cargo check with unstable-cloud and rustls-x features
+        run: cargo check --all-targets --manifest-path "scylla/Cargo.toml" --features "unstable-cloud" --features "rustls-023" 
+      - name: Cargo check with unstable-cloud, openssl-x and rustls-x features
+        run: cargo check --all-targets --manifest-path "scylla/Cargo.toml" --features "unstable-cloud" --features "openssl-010" --features "rustls-023" 
```
:pick: trailing whitespace (visible in `git diff`).

Lorak-mmk reviewed (approved):
@muzarski Please fix the conflict and we can merge!

I'm awaiting the day we will be able to bump MSRV to a version that supports resolver v3 and drop Cargo.lock.msrv. About 6  months left.

muzarski commented:
> Doesn't this PR also close #771 ?

REading the discussion from the issue, I believe that this PR closes it.

muzarski commented:
> > Doesn't this PR also close #771 ?
> 
> After rEading the discussion from the issue, I believe that this PR closes it.
